File: 2022/week_1/psets/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

char check_validity(long card);

int main(void)
{
    // Getting user input
    long card = get_long("Number: ");

    // Determining card type and checking validity
    if (check_validity(card))
    {
        if (card / (long) pow(10, 13) == 34 || card / (long) pow(10, 13) == 37)
        {
            printf("AMEX\n");
        }
        else if (card / (long) pow(10, 14) == 51 ||
                 card / (long) pow(10, 14) == 52 ||
                 card / (long) pow(10, 14) == 53 ||
                 card / (long) pow(10, 14) == 54 ||
                 card / (long) pow(10, 14) == 55
                )
        {
            printf("MASTERCARD\n");
        }
        else if (card / (long) pow(10, 12) == 4 || card / (long) pow(10, 15) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

char check_validity(long card)
{
    // Luhn's algorithm
    int sum = 0;
    int num;

    int i = 0;
    do
    {
        i += 2;
        num = ((card % (long) pow(10, i)) / (long) pow(10, i - 1)) * 2;
        sum += (num / 10 > 0) ? ((num / 10) + (num % 10)) : num;
    }
    while (card / (long) pow(10, i) > 0);

    i = -1;
    do
    {
        i += 2;
        sum += (card % (long) pow(10, i) / (long) pow(10, i - 1));
    }
    while (card / (long) pow(10, i) > 0);

    return (sum % 10 == 0) ? 1 : 0;
}

File: 2022/week_6/psets/sentimental-credit/credit.py
# Credit pset in python
import re


def main():
    # Getting user input
    num_regex = re.compile("\d*")
    try:
        num = input("Number: ")
        i = int(num)
    except ValueError:
        pass

    # Compiling regex validators
    visa = re.compile("^4(\d{12}|\d{15})$")
    mc = re.compile("^(5[1-5])+\d{14}$")
    amex = re.compile("^(34|37)+\d{13}$")

    # Checking validity
    if amex.match(num) and check_validity(num):
        print("AMEX")

    elif visa.match(num) and check_validity(num):
        print("VISA")

    elif mc.match(num) and check_validity(num):
        print("MASTERCARD")

    else:
        print("INVALID")


def check_validity(card_number):
    '''
    Using Luhn’s Algorithm to check the validity of the card number
    '''
    number = card_number[::-1]
    every_other = [int(number[i]) * 2 for i in range(1, len(number), 2)]
    total = sum([n if n // 10 == 0 else (n // 10 + n % 10) for n in every_other])
    total += sum([int(number[i]) for i in range(0, len(number), 2)])
    return True if total % 10 == 0 else False


if __name__ == "__main__":
    main()